python/lsst/cell_coadds/_identifiers.py

```diff
@@ -53,17 +55,37 @@ class PatchIdentifiers:
     """Identifiers for the patch itself.
     """
 
-    band: Optional[str]
+    band: str | None
     """Name of the band, if any.
     """
 
     @classmethod
-    def from_data_id(cls, data_id: DataCoordinate) -> PatchIdentifiers:
+    def from_data_id(cls: Type[T], data_id: DataCoordinate, **kwargs: Any) -> T:
+        """Construct from a data ID."""
+        # The 'type: ignore' directives below are present because we require a
+        # a fully-expanded data ID, but this isn't embedded in the types.
```

Comment:
Maybe the `.records` ones are, but at least some of them are actually saying "this might be `str` or it might be `int`! Data IDs can have both!", which is of course also something we should ignore or `typing.cast` (no strong preference for which in this case).